### PR TITLE
Event list

### PR DESCRIPTION
**Pull request by sam-brett**
Still adding tests, but the code can be reviewed as it is pretty static

### CODE REVIEW

**Comment by sam-brett** [comment]
@vinopm I've made a fairly dirty merge here, but its allows it to merge automatically now. There is one test that has broken in the Isolation policy area (testShouldIsolate_UsesIndividualPolicy_InfectedAndNoneTestedPositive) I think this might be to do with the infection schema change.

File: src/main/java/uk/co/ramp/event/types/FormattedEvent.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@JsonSerialize
+@JsonDeserialize
+@JsonPropertyOrder({"time", "eventType", "id", "newStatus", "additional"})
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/types/FormattedEvent.java]:
```suggestion
@JsonPropertyOrder({"time", "eventType", "id", "newStatus", "additionalInfo"})
```

File: src/main/java/uk/co/ramp/io/ContactReader.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import java.io.IOException;
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Service
+public class ContactReader {
+    private static final Logger LOGGER = LogManager.getLogger(ContactReader.class);
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/io/ContactReader.java]: Unused: 
```suggestion
```

File: src/main/java/uk/co/ramp/io/ContactReader.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    public Map<Integer, List<ContactEvent>> readEvents(Reader reader, StandardProperties properties) throws IOException {
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/io/ContactReader.java]:
```suggestion
    private final StandardProperties properties;
    @Autowired
    public ContactReader(StandardProperties standardProperties) {
        this.properties = standardProperties;
    }

    public Map<Integer, List<ContactEvent>> readEvents(Reader reader) throws IOException {
```
We could avoid passing in standardproperties (which is pretty much static, does not change per method call) by passing it through constructor and injecting it via Spring

File: src/main/java/uk/co/ramp/io/ContactReader.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            if (event.from() >= properties.populationSize() ||
+                    event.to() >= properties.populationSize() ||
+                    event.time() > properties.timeLimit()) continue;
+
+            dailyRecord.putIfAbsent(event.time(), new ArrayList<>());
+            dailyRecord.get(event.time()).add(event);
+
+        }
+
+        return dailyRecord;
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/io/ContactReader.java]:
```suggestion
return contactEvents.stream()
                .filter(event -> event.from() < properties.populationSize())
                .filter(event -> event.to() < properties.populationSize())
                .filter(event -> event.time() <= properties.timeLimit())
                .collect(Collectors.groupingBy(ContactEvent::time));
```

File: src/main/java/uk/co/ramp/people/AlertStatus.java [hunk 5]
```diff
@@ -26,7 +27,8 @@
         validTransitions.put(ALERTED, List.of(REQUESTED_TEST));
         validTransitions.put(REQUESTED_TEST, List.of(AWAITING_RESULT));
         //TODO: get clarification here
-        validTransitions.put(AWAITING_RESULT, List.of(TESTED_POSITIVE, ALERTED, NONE));
+        validTransitions.put(AWAITING_RESULT, List.of(TESTED_POSITIVE, TESTED_NEGATIVE));
+        validTransitions.put(TESTED_NEGATIVE, List.of(NONE));
         validTransitions.put(TESTED_POSITIVE, Collections.emptyList());
     }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/people/AlertStatus.java:34]: transitionTo() not used anymore. we can remove this?

File: src/main/java/uk/co/ramp/people/Case.java [hunk 6]
```diff
@@ -1,156 +1,133 @@
...
 
-    public void setStatus(VirusStatus newStatus) {
-        this.status = this.status.transitionTo(newStatus);
-    }
+    //derived values
 
-    public int nextAlertStatusChange() {
-        return nextAlertStatusChange;
+    public String getSource() {
+        return Strings.padEnd(id() + "(" + exposedTime + ")", 12, ' ');
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/people/Case.java]: The padding should probably live within the implementation of InfectionMap rather than Case? id and exposedTime can already be accessed from InfectionMap

File: src/main/java/uk/co/ramp/people/Case.java [hunk 7]
```diff
@@ -1,156 +1,133 @@
...
-    // statics
+        if (event instanceof VirusEvent) {
+            processVirusEvent((VirusEvent) event);
+        } else if (event instanceof InfectionEvent) {
+            processInfectionEvent((InfectionEvent) event);
+        } else if (event instanceof AlertEvent) {
+            processAlertEvent((AlertEvent) event);
+        } else {
+            throw new EventException("uncovered condition");
+        }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/people/Case.java]: Relying on `instanceof` instead of polymorphism is generally considered code smell within OOP. We could possibly resolve this by having an interface method in Event called something like applyEventToCase(Case aCase) and the children of Event such as VirusEvent/InfectionEvent/AlertEvent to have their own implementations for applyEventToCase.

For example the VirusEvent interface would have:
```
default void applyEventToCase(Case aCase) {
    aCase.setVirusStatus(aCase.getVirusStatus().transitionTo(newStatus());
}
```
or 
```
public void applyEventToCase(Case aCase) {
    aCase.setVirusStatus(aCase.getVirusStatus().transitionTo(newStatus());
}
```

and then in EventProcessor you would have 
`event.applyEventToCase(population.get(event.id())` instead of `population.get(event.id()).processEvent(event)`

File: src/test/java/uk/co/ramp/policy/SingleCaseIsolationPolicyTest.java [hunk 8]
```diff
@@ -132,22 +128,24 @@ public void setUp() {
...
-        assertThat(isolationPolicy.isIndividualInIsolation(id, INFECTED_SYMP, TESTED_POSITIVE, compliance, proportionOfPopulationInfected, currentTime)).isTrue();
+        assertThat(isolationPolicy.isIndividualInIsolation(id, SYMPTOMATIC, TESTED_POSITIVE, compliance, proportionOfPopulationInfected, currentTime)).isTrue();
     }
 
     @Test
+    @Ignore
+    //TODO fix
     public void testShouldIsolate_UsesIndividualPolicy_InfectedAndNoneTestedPositive() {
-        assertThat(isolationPolicy.isIndividualInIsolation(id, INFECTED, NONE, compliance, proportionOfPopulationInfected, currentTime)).isFalse();
+        assertThat(isolationPolicy.isIndividualInIsolation(id, SYMPTOMATIC, NONE, compliance, proportionOfPopulationInfected, currentTime)).isFalse();
```
Comment (vinopm) [inline src/test/java/uk/co/ramp/policy/SingleCaseIsolationPolicyTest.java]:
```suggestion
        assertThat(isolationPolicy.isIndividualInIsolation(id, ASYMPTOMATIC, NONE, compliance, proportionOfPopulationInfected, currentTime)).isFalse();
```

This should be ASYMPTOMATIC instead. Perhaps the test name should be updated too: testShouldIsolate_UsesIndividualPolicy_AsymptomaticAndNoAlertStatus.

File: src/test/java/uk/co/ramp/policy/SingleCaseIsolationPolicyTest.java [hunk 9]
```diff
@@ -132,22 +128,24 @@ public void setUp() {
 
     @Test
     public void testShouldIsolate_UsesIndividualPolicy_InfectedAndTestedPositive() {
-        assertThat(isolationPolicy.isIndividualInIsolation(id, INFECTED_SYMP, TESTED_POSITIVE, compliance, proportionOfPopulationInfected, currentTime)).isTrue();
+        assertThat(isolationPolicy.isIndividualInIsolation(id, SYMPTOMATIC, TESTED_POSITIVE, compliance, proportionOfPopulationInfected, currentTime)).isTrue();
     }
 
     @Test
+    @Ignore
+    //TODO fix
```
Comment (vinopm) [inline src/test/java/uk/co/ramp/policy/SingleCaseIsolationPolicyTest.java]:
```suggestion

```

File: src/main/java/uk/co/ramp/event/EventList.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    public void addEvents(Map<Integer, List<ContactEvent>> readEvents) {
+
+        for (Map.Entry<Integer, List<ContactEvent>> entry : readEvents.entrySet()) {
+            bulkAdd(entry.getKey(), entry.getValue());
+        }
+    }
+
+    private void bulkAdd(int time, List<ContactEvent> events) {
+        map.computeIfAbsent(time, k -> new ArrayList<>()).addAll(events);
+    }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventList.java]:
```suggestion
    public void addEvents(Map<Integer, List<ContactEvent>> readEvents) {
        readEvents.values().stream().flatMap(List::stream).forEach(this::addEvent);
    }
```

File: src/main/java/uk/co/ramp/event/EventList.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+
+    //read
+    public List<Event> getForTime(int time) {
+        return map.getOrDefault(time, new ArrayList<>());
+    }
+
+    public Map<Integer, List<Event>> getMap() {
+        return map;
+    }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventList.java]: It is unnecessary to expose the map publicly. The only caller of this is from outbreak where outbreak.java is computing the last contact. We can replace this method with the following:
```suggestion
    public Optional<Integer> lastContactTime() {
        return map.keySet().stream().max(Comparator.naturalOrder());
    }
```

and in Outbreak.java we can call:
`int lastContact = eventList.lastContactTime().orElseThrow();`
instead of 
`int lastContact = eventList.getMap().keySet().stream().max(Comparator.naturalOrder()).orElseThrow();`

File: src/main/java/uk/co/ramp/event/EventList.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+        map.computeIfAbsent(time, k -> new ArrayList<>()).addAll(events);
+    }
+
+    public void completed(Event e) {
+        completedMap.computeIfAbsent(e.time(), k -> new ArrayList<>()).add(e);
+    }
+
+
+    //read
+    public List<Event> getForTime(int time) {
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventList.java]: This can be package-private to avoid the possibility of code outside of the event package being able to modify the list
```suggestion
   List<Event> getForTime(int time) {
```

File: src/main/java/uk/co/ramp/event/EventList.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        try (Writer writer = new FileWriter("events.csv")) {
+            new CsvWriter().write(writer, finalList, ImmutableFormattedEvent.class);
+
+        } catch (IOException e) {
+            String message = "An error occurred while writing a CSV file";
+            LOGGER.error(message);
+            throw new CsvException(message, e);
+        }
+
+    }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventList.java]: This could be moved to an EventOutputWriter.java to separate responsibilities

File: src/main/java/uk/co/ramp/event/EventList.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        return map;
+    }
+
+    public void output() {
+
+        List<Integer> timeStamps = new ArrayList<>(completedMap.keySet());
+        timeStamps.sort(Comparator.naturalOrder());
+        List<ImmutableFormattedEvent> finalList = timeStamps.stream().map(completedMap::get).flatMap(Collection::stream).map(formattedEventFactory::create).filter(Objects::nonNull).collect(Collectors.toList());
+
+        try (Writer writer = new FileWriter("events.csv")) {
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventList.java]: Pass in 'event.csv' file name via config or declare it as a static class variable

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+        this.utils = utils;
+        this.eventList = eventList;
+        this.diseaseProperties = diseaseProperties;
+        this.isolationPolicy = isolationPolicy;
+    }
+
+
+    public void setPopulation(Map<Integer, Case> population) {
+        this.population = population;
+    }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java:54]: Would it be possible to pass the population object in through the constructor instead?

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+
+        }
+        return Optional.empty();
+    }
+
+    private AlertStatus determineNextAlertStatus(AlertEvent event) {
+
+        Case thisCase = population.get(event.id());
+
+        List<AlertStatus> newStatus = getValidTransitions(event.newStatus());
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java]: Perhaps calling newStatus as nextStatus would be more readable

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+        switch (event.newStatus()) {
+            case AWAITING_RESULT:
+                return thisCase.isInfectious() ? TESTED_POSITIVE : TESTED_NEGATIVE;
+            case NONE:
+                return NONE;
+            case TESTED_POSITIVE:
+                return TESTED_POSITIVE;
+            default:
+                LOGGER.error(event);
+                throw new EventException("There is no case for the event" + event);
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java]: Might be more clear if we call newStatus as nextStatus instead.

If the event.newStatus() is NONE, the nextstatus (based on valid transitions) are a list of [ALERTED, REQUESTED_TEST], but here we will always return NONE? (line 260)

Comment (vinopm) [reply src/main/java/uk/co/ramp/event/EventProcessor.java]: Also I think the case TESTED_POSITIVE is not reachable because of line 253

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+            AlertStatus nextStatus = determineNextAlertStatus(event);
+
+            if (nextStatus != event.newStatus()) {
+                int nextTime = timeInStatus(nextStatus);
+
+                AlertEvent subsequentEvent = ImmutableAlertEvent.builder().
+                        id(event.id()).
+                        oldStatus(event.newStatus()).
+                        newStatus(nextStatus).
+                        time(time + nextTime).
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java]: confusing to have it called nextTime, may better to call it nextTimeDelta?

File: src/main/java/uk/co/ramp/event/EventProcessor.java (hunk 17 above)
Comment (vinopm) [reply src/main/java/uk/co/ramp/event/EventProcessor.java]: Also this skips the validity checking we have in place to check if the next transition is valid

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+        if (newStatus.size() == 1) return newStatus.get(0);
+
+        // these cases have multiple paths
+        switch (event.newStatus()) {
+            case EXPOSED:
+                return determineInfection(population.get(event.id()));
+            case SYMPTOMATIC:
+                return determineSeverity(population.get(event.id()));
+            case SEVERELY_SYMPTOMATIC:
+                return determineOutcome(population.get(event.id()));
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java]: This skips the validity checking that we have in place to check if the next transition is valid

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+        // Look at all virus events
+        for (VirusEvent event : eventList.getForTime(time).stream().filter(event -> event instanceof VirusEvent).map(event -> (VirusEvent) event).collect(Collectors.toList())) {
+
+            population.get(event.id()).processEvent(event);
+            VirusStatus nextStatus = determineNextStatus(event);
+
+            // will return self if at DEAD or RECOVERED
+            if (event.newStatus() != nextStatus) {
+
+                int nextTime = timeInCompartment(event.newStatus(), nextStatus);
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java]:
```suggestion
                int nextTimeDelta = timeInCompartment(event.newStatus(), nextStatus);
```

File: src/main/java/uk/co/ramp/event/EventProcessor.java [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+
+                VirusEvent subsequentEvent = ImmutableVirusEvent.builder().
+                        id(event.id()).
+                        oldStatus(event.newStatus()).
+                        newStatus(nextStatus).
+                        time(time + nextTime).
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventProcessor.java]:
```suggestion
                        time(time + nextTimeDelta).
```

File: src/main/java/uk/co/ramp/event/FormattedEventFactory.java [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+import uk.co.ramp.people.Case;
+
+@Service
+public class FormattedEventFactory {
+
+    private static final Logger LOGGER = LogManager.getLogger(FormattedEventFactory.class);
+
+    public ImmutableFormattedEvent create(Event event) {
+
+        if (event instanceof VirusEvent) {
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/FormattedEventFactory.java:16]: The use of instanceof here seems like code smell. We may have to refactor but probably best to leave that for a separate PR as I think its not a simple fix

File: src/main/java/uk/co/ramp/event/FormattedEventFactory.java [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+                eventType("AlertEvent").
+                id(event.id()).
+                newStatus(event.newStatus().toString()).
+                additionalInfo("").
+                build();
+    }
+
+
+    ImmutableFormattedEvent create(PolicyEvent event) {
+        // TODO: check when merging
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/FormattedEventFactory.java]: Should be fine to leave this as it is for now

File: src/test/java/uk/co/ramp/event/FormattedEventFactoryTest.java [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+        Assert.assertNotNull(event.eventType());
+        Assert.assertEquals("PolicyEvent", event.eventType());
+
+        try {
+            formattedEventFactory.create(notCovered);
+
+        } catch (EventException e) {
+            Assert.assertThat(logSpy.getOutput(), containsString(notCovered.getClass().getSimpleName()));
+            throw e;
+        }
```
Comment (vinopm) [inline src/test/java/uk/co/ramp/event/FormattedEventFactoryTest.java]:
```suggestion
        assertThatExceptionOfType(EventException.class)
                .isThrownBy(() -> formattedEventFactory.create(notCovered))
                .withMessageContaining("Unknown Event Type")
                .withMessageContaining(notCovered.getClass().getSimpleName());
```

File: src/main/java/uk/co/ramp/Outbreak.java [hunk 25]
```diff
@@ -63,171 +70,71 @@ public void setContactRecords(Map<Integer, List<ContactRecord>> contactRecords)
...
-
-    int calculateDailyStatistics(int time) {
+    void updateLogActiveCases(int time) {
+        int previousActiveCases = activeCases;
         Map<VirusStatus, Integer> stats = utils.getCmptCounts(population);
-        int activeCases = stats.get(EXPOSED) + stats.get(EXPOSED_2) + stats.get(INFECTED) + stats.get(INFECTED_SYMP);
+        activeCases = stats.get(EXPOSED) + stats.get(ASYMPTOMATIC) + stats.get(PRESYMPTOMATIC) + stats.get(SYMPTOMATIC) + stats.get(SEVERELY_SYMPTOMATIC);
 
-        CmptRecord cmptRecord = outputLog.log(time, stats);
+        CmptRecord cmptRecord = outputLog.log(time, stats, activeCases - previousActiveCases);
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/Outbreak.java]: I would push storing the state of activeCases into the implementation of LogDailyOutput. This would mean we can simply pass in (time, stats) as we were doing before, and let the LogDailyOutput class keep track of the previous cases.

File: src/main/java/uk/co/ramp/Outbreak.java [hunk 26]
```diff
@@ -63,171 +70,71 @@ public void setContactRecords(Map<Integer, List<ContactRecord>> contactRecords)
...
-                    compartmentCounts.get(INFECTED) == 0 &&
-                    compartmentCounts.get(INFECTED_SYMP) == 0;
-
-
-            if (exitCondition) {
-                LOGGER.info("Steady state solution reached at t={} ", time);
-                LOGGER.info("Exiting early.");
-                return;
+            // stop random infections after contacts end
+            if (activeCases == 0 && lastContact < time) {
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/Outbreak.java]: I would calculate this on the fly instead of storing state in class.

Perhaps a separate private method to call from here:
```
    private int activeCases() {
        var stats = utils.getCmptCounts(population);
        return stats.get(EXPOSED) + stats.get(ASYMPTOMATIC) + stats.get(PRESYMPTOMATIC) + stats.get(SYMPTOMATIC) + stats.get(SEVERELY_SYMPTOMATIC);
    }
```

File: src/main/java/uk/co/ramp/Outbreak.java [hunk 27]
```diff
@@ -63,171 +70,71 @@ public void setContactRecords(Map<Integer, List<ContactRecord>> contactRecords)
...
+                exposedBy(Case.getInitial()).
+                oldStatus(SUSCEPTIBLE).
+                newStatus(EXPOSED).
+                exposedTime(0).
+                id(-1).
+                time(0).build();
 
-            EvaluateCase evaluateCase = new EvaluateCase(population.get(id), diseaseProperties, distributionSampler);
-            evaluateCase.updateVirusStatus(EXPOSED, 0, Case.getInitial());
+        infectedIds.forEach(id -> virusEvents.add(ImmutableInfectionEvent.builder().from(genericEvent).id(id).build()));
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/Outbreak.java]: Personal preference but feel free to ignore:
```suggestion
        infectedIds.forEach(id -> virusEvents.add(ImmutableInfectionEvent.copyOf(genericEvent).withId(id)));
```

**Review by vinopm — COMMENTED** [review]
@sam-brett I have reviewed all changes so far and left some suggestions/comments. Looks good to me on the whole, nice work with the event list. There are some uses of `instanceof` that I'd like us to avoid using, but perhaps we can address that in a follow-up PR. I'll review this PR again once the rest of your tests are in (as you mentioned previously).

File: src/main/java/uk/co/ramp/Outbreak.java (hunk 26 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/Outbreak.java]: Fair point

File: src/main/java/uk/co/ramp/Outbreak.java (hunk 25 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/Outbreak.java]: Yep that makes sense, although it does mean we have active cases being calculated in two places, but I don't think that matters.

File: src/main/java/uk/co/ramp/event/EventList.java (hunk 11 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventList.java]: good call

File: src/main/java/uk/co/ramp/event/EventList.java (hunk 14 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventList.java]: I've made it static for now, but it might be worth making an output config at some point so the output files are defined externally should the user want to pipe them elsewhere.

File: src/main/java/uk/co/ramp/Outbreak.java (hunk 25 above)
Comment (vinopm) [reply src/main/java/uk/co/ramp/Outbreak.java]: Nice. Yes you're right. Seems like a premature optimisation to me to calculate once and pass value through. My thinking was that this suggestion would simplify the code from a readability perspective

File: src/main/java/uk/co/ramp/event/EventProcessor.java (hunk 15 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventProcessor.java:54]: I guess we could, I think we'd need to make the population a bean first though.

File: src/main/java/uk/co/ramp/event/EventProcessor.java (hunk 18 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventProcessor.java]: changed to deltaTime

File: src/main/java/uk/co/ramp/event/EventProcessor.java (hunk 16 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventProcessor.java]: done

File: src/main/java/uk/co/ramp/event/EventProcessor.java (hunk 19 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventProcessor.java]: Added logic to the three methods

File: src/main/java/uk/co/ramp/event/FormattedEventFactory.java (hunk 22 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/FormattedEventFactory.java:16]: Yeah I winced when I was writing it.

File: src/main/java/uk/co/ramp/people/Case.java (hunk 6 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/people/Case.java]: moved

File: src/main/java/uk/co/ramp/people/AlertStatus.java (hunk 5 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/people/AlertStatus.java:34]: oops. I added the validity check back into even processor.

**Comment by sam-brett** [comment]
@vinopm I think that covers everything except the instanceOf calls.

File: src/test/java/uk/co/ramp/io/InfectionMapTest.java [hunk 28]
```diff
@@ -1,14 +1,152 @@
...
+        infectionMap = mock(InfectionMap.class);
+        doCallRealMethod().when(infectionMap).outputMap(any());
+        doThrow(new IOException("")).when(infectionMap).recurseSet(anyList(), anyMap(), any(), anyInt());
+
+        try {
+            infectionMap.outputMap(new StringWriter());
+        } catch (InfectionMapException e) {
+            Assert.assertThat(logSpy.getOutput(), containsString("An error occurred while writing the map file"));
+            throw e;
+        }
```
Comment (vinopm) [inline src/test/java/uk/co/ramp/io/InfectionMapTest.java]:
```suggestion
assertThatExceptionOfType(InfectionMapException.class)
                .isThrownBy(() -> infectionMap.outputMap(new StringWriter()))
                .withMessageContaining("An error occurred while writing the map file");
```

File: src/test/java/uk/co/ramp/io/InfectionMapTest.java [hunk 29]
```diff
@@ -1,14 +1,152 @@
...
+                stringWriter.toString().split("\n")).
+                filter(s -> !s.isBlank()).
+                map(s -> s.substring(0, s.indexOf("("))).
+                map(Integer::parseInt).
+                collect(Collectors.toSet());
+
+        Assert.assertEquals(lines, seeds);
+    }
+
+    @Test(expected = InfectionMapException.class)
```
Comment (vinopm) [inline src/test/java/uk/co/ramp/io/InfectionMapTest.java]:
```suggestion
```

Not needed if you use `assertThatExceptionOfType()`

File: src/main/java/uk/co/ramp/io/InfectionMap.java [hunk 30]
```diff
@@ -89,5 +72,29 @@ void recurseSet(List<Case> target, Map<Integer, Set<Case>> infectors, Writer wri
 
     }
 
+    Map<Integer, Set<Case>> collectInfectors() {
+
+        List<Case> infections = population.values().stream()
+                .filter(c -> c.virusStatus() != SUSCEPTIBLE)
+                .sorted(Comparator.comparingInt(Case::exposedTime))
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/io/InfectionMap.java:79]: I couldn't work out what the purpose of sorting this here is. We seem to be returning a set rather than a list?

File: src/main/java/uk/co/ramp/event/EventList.java [hunk 31]
```diff
@@ -0,0 +1,89 @@
...
+
+
+    //read
+    List<Event> getForTime(int time) {
+        return map.getOrDefault(time, new ArrayList<>());
+    }
+
+    public Map<Integer, List<Event>> getMap() {
+        return map;
+    }
```
Comment (vinopm) [inline src/main/java/uk/co/ramp/event/EventList.java]: This can be removed now:
```suggestion

```

**Review by vinopm — COMMENTED** [review]
@sam-brett Thanks. I've marked the ones you have addressed as resolved. There are still 8 or so open comments/suggestions excluding the instanceof comment. Let me know what your thoughts are on it

**Comment by vinopm** [comment]
@sam-brett There are still 8 or so open comments/suggestions excluding the instanceof comment. Let me know what your thoughts are on it

File: src/main/java/uk/co/ramp/event/EventProcessor.java (hunk 17 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/event/EventProcessor.java]: I'll remove that condition, as you're right about it being returned earlier. The transition is checked at line 116.

File: src/main/java/uk/co/ramp/io/InfectionMap.java (hunk 30 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/io/InfectionMap.java:79]: Replaced with list. This was to make sure the infections printed in time order, but obviously the set doesn't keep that so good spot.

File: src/main/java/uk/co/ramp/people/Case.java (hunk 7 above)
Comment (sam-brett) [reply src/main/java/uk/co/ramp/people/Case.java]: Great spot. changed.

**Review by vinopm — APPROVED** [review]
Approved
